File: 208. 实现 Trie (前缀树).py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.is_end = False


class Trie(object):

    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        """
        :type word: str
        :rtype: None
        """
        curr = self.root
        for char in word:
            if char not in curr.children:
                curr.children[char] = TrieNode()
            curr = curr.children[char]
        curr.is_end = True

    def search(self, word):
        """
        :type word: str
        :rtype: bool
        """
        curr = self.root
        for char in word:
            if char not in curr.children:
                return False
            curr = curr.children[char]
        # if curr.is_end==False:
        #     return False
        # return True
        return curr.is_end

    def startsWith(self, prefix):
        """
        :type prefix: str
        :rtype: bool
        """
        curr = self.root
        for char in prefix:
            if char not in curr.children:
                return False
            curr = curr.children[char]
        return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: 208. 实现 Trie (前缀树).go
package main

// 记得传指针

type Node struct {
	Children map[byte]*Node
	isEnd    bool
}
type Trie struct {
	root *Node
}

func Constructor() Trie {
	return Trie{
		root: &Node{
			Children: make(map[byte]*Node),
			isEnd:    false,
		},
	}
}

func (this *Trie) Insert(word string) {
	curr := this.root
	for i := range len(word) {
		if _, ok := curr.Children[word[i]]; !ok {
			curr.Children[word[i]] = &Node{Children: make(map[byte]*Node)}
		}
		curr = curr.Children[word[i]]
	}
	curr.isEnd = true
}

func (this *Trie) Search(word string) bool {
	curr := this.root
	for i := range len(word) {
		if _, ok := curr.Children[word[i]]; !ok {
			return false
		}
		curr = curr.Children[word[i]]
	}
	if curr.isEnd != true {
		return false
	}
	return true
}

func (this *Trie) StartsWith(prefix string) bool {
	curr := this.root
	for i := range len(prefix) {
		if _, ok := curr.Children[prefix[i]]; !ok {
			return false
		}
		curr = curr.Children[prefix[i]]
	}
	return true
}

/**
 * Your Trie object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Insert(word);
 * param_2 := obj.Search(word);
 * param_3 := obj.StartsWith(prefix);
 */
